### ISSUE TITLE
[Overview] Upgrade to new Framework and Tempate System modules

### ISSUE THREAD

**Issue by eliot-akira** [TangibleInc/template-system#16]
This document tracks the progress of migrating from the deprecated Interface module and Plugin Framework v2, to the new modules in the Framework and Tempate System.

See documentation pages for:

- [Template System: Modules](https://docs.loopsandlogic.com/reference/template-system/modules/)
- [Framework module](https://docs.loopsandlogic.com/reference/framework/)

This will be relevant for plugins as they move over to using the new framework.

- [ ] Document upgrade process

### Plugin framework

Here are the features that were in the old plugin framework. They're being moved and rewritten into new modules.

- [x] AJAX → `framework/ajax`
- [x] Date module → `framework/date`
- [x] HJSON module → `framework/hjson`
- [x] HTML → `framework/html`
- [ ] Preact → `framework/preact`
- [ ] Utilities → `framework/format`, `log`

Server-side utilities

- [ ] Async Action
- [ ] Background Queue
- [ ] File System

#### Plugin features

- [ ] Settings page with tabs
- [ ] Support page
- [ ] Plugin dependency check 
- [x] Plugin updater → [Plugin Updater module](https://bitbucket.org/tangibleinc/plugin-updater)
- [ ] Licensing - Was using EDD license server

#### Post type extensions
- [x] Sortable post type → `admin/post-types/sortable-post-type`
- [x] Duplicate posts → `admin/post-types/duplicate-post`

### Interface module

Here are the features that were in the Interface module. These are now consolidated with other Template System modules.

- [ ] Chart
- [ ] Date picker
- [ ] Embed
- [ ] Glider
- [ ] Prism
- [ ] Select
- [ ] Slider
- [ ] Sortable
- [ ] Table (sortable and paginated)

**Comment by GabrielGallagher** [comment]
Hi Eliot, 

We have a lot of in-development (or released) plugins on the old framework. 
Should we be switching to the new version of the framework for those?
I guess a lot of them might rely on the dependency check and background queue stuff, so maybe we have to wait?

**Comment by eliot-akira** [comment]
> We have a lot of in-development (or released) plugins on the old framework. Should we be switching to the new version of the framework for those?

Yes, it would be good to gradually migrate these plugins to use the new framework, since it's slimmer and will be updated going forward. But it's not urgent, since the old plugin framework can co-exist with it and will continue to work.

> I guess a lot of them might rely on the dependency check and background queue stuff, so maybe we have to wait?

I think we can start upgrading the plugins, and move over any remaining features from the old plugin framework as needed.

As part of the upgrade, I'd like to move them to public or private repositories on GitHub.

---

A question to be solved still is how to handle plugin updates. Currently, there are several ways plugins are updated:

- L&L: Published and updated via WordPress repository
- Tangible Blocks/Pro/Editor (and maybe others): Published by Bitbucket pipeline (v2) to the plugin update server (`updater.tangible.one`), and updated by the [Plugin Updater module](https://bitbucket.org/tangibleinc/plugin-updater)
- Older plugins (?): Published by Bitbucket pipeline (v1) to the EDD store, and updated by the EDD plugin updater in the old plugin framework

This could be greatly simplified by using:

- [GitHub Actions](https://docs.github.com/en/actions) to prepare releases
  Build zip packages on every commit and version tag. Maybe publish to Cloudflare.
- [Git Updater](https://github.com/afragen/git-updater/) to update plugins directly from GitHub

For private repos, such as the Pro plugins, we'll still need our own plugin updater and update server for licensing.
